---
title: "An Introduction to the multiGSEA Package"
author: "Steve Lianoglou"
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('multiGSEA')`"
abstract: >
  The multiGSEA package facilitates the execution and comparison of several
  GSEA approaches for a single experimental contrast and provides tools
  to explore these results from within the user's workspace or via a shiny
  application.
vignette: >
  %\VignetteIndexEntry{An Introduction to the multiGSEA Package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding[utf8]{inputenc}
output:
  html_vignette:
    toc: true
    toc_depth: 4
    toc_float: 
      collapsed: false
      smooth_scroll: yes
    css: mystyles.css
    self_contained: yes
---

```{r init, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
library(BiocStyle)
library(multiGSEA)
knitr::opts_chunk$set(
  echo=TRUE, warning=FALSE, message=FALSE, error=FALSE) #, dpi=150)
```

# Overview

The `multiGSEA` package was built to facilitate the use of gene sets in the
analysis of high throughput genomics data (primarily RNA-seq). It does so
by providing these top-line functionalities:

* The `multiGESA` function is a wrapper that orchestrates the execution of any
  number of user-specified gene set enrichment analyses (GSEA) over a particular
  experimental contrast of interest. This will create a `MultiGSEAResult`
  object which stores the resutls of each GSEA method internally, allowing
  for easy query and retrieval.
* The `explore` function launches a shiny application that facilitates the
  interactive exploration of these GSEA results. This application lives in
  `inst/shiny` and can also be deployed to a shiny server and can be initialized
  by uploading a serialized MultiGSEAResult *.rds file.
* The `scoreSingeSamples` function is a wrapper that enables the user to
  generate single sample gene set scores using a variety of different
  single sample gene set scoring methods.
* The `GeneSetDb` class: a new class to store collections of gene sets. This
  provides a bit more functionality than the base `GeneSetCollection` class
  defined in the [GSEABase][http://bioconductor.org/packages/GSEABase].

The initial GSEA methods that multiGSEA wrapped were the ones provided by limma
and edgeR. As such, many analyses using multiGSEA expect you to re-use the same
data objects used for differential expression analysis, namely:

* Expression data (an `EList`, `DGEList`, or expression matrix)
* A design matrix
* A contrast vector/matrix (if your design and comparison require it)

Other methods only require the user to provide a vector of statistics that
represent some differential expression statistic per gene, and the GSEA is
performed by analyzing the ranks of genes within this vector.

All GSEA methods require the use of a `GeneSetDb`.

Currenty supported gene set enrichment methods include:

```{r multiGSEA-methods, results='asis', echo=FALSE}
cat(sprintf('* %s\n', multiGSEA.methods()))
```

When using these methods in analyses that lead to publication, please cite
the original papers that developed these methods and cite multiGSEA when its
functionality assisted in your interpretation and analysis.

# Standard Workflow

We'll use the `"r Biocpkg("airway")"` data package from [Himes et al.][himes] as
the basis for our analysis here[^rnaseqworkflow]. This dataset contains
treatment (with the glucocorticoid dexamethasone) and control samples from four
airway smooth muscle (ASM) cell lines in order to study the mechanism by which
glucocorticoids suppress inflammation in ASM.

[himes]: http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0099625

[^rnaseqworkflow]: This dataset is also used as the basis for the RNA-seq
workflow tutorial published by Love et al. on
[f1000research](http://f1000research.com/articles/4-1070/v1)

Let's setup our work environment for analysis with multiGSEA.

```{r init-env}
library(multiGSEA)
library(magrittr)
library(dplyr)
library(ggplot2)
theme_set(theme_bw())
options(multiGSEA.df.return='data.frame')
```

Internally, mulitGSEA leverages the
[`data.table`][https://CRAN.R-project.org/package=data.table] package for fast
indexing and manipulation over data.frames. There are many functions that return
data.tables back to the end user, but some users may not want to work with those.
The `multiGSEA.df.return` option can be set to `data.frame` (as above) so that
all data.frame-like objects returned by this package are actually `data.frames`,
and not `data.tables`.

## Data Setup

multiGSEA is most straightforward to use when our data objects and analysis are
performed with either the edgeR or voom/limma pipelines. In order to facilitate
our analyses, we first convert the `airway` `SummarizedExperiment` to a
`DGEList` and add simple gene annotations to the data.

We provide a `createAirwayDGEList` function that encapsulates the logic
described above and returns a DGEList from the airway dataset with Entrez
identifiers as its gene-level identifiers.

```{r data-setup, eval=!exists('y.all'), results='hide'}
y.all <- createAirwayDGEList()
```

Here is an overview of the data we are working with

`r knitr::kable(y.all$samples)`

## Data Analysis

We want to identify the genes (and genesets) regulated under `dex` treatment.

We'll use the following design:

```{r design-setup}
des <- model.matrix(~ 0 + group + cell, y.all$samples)
colnames(des) <- sub('group', '', colnames(des))
knitr::kable(as.data.frame(des))
```

To test for the general affect of dexamethasone treatment independant of
cell-type, we would test the following contrast:

```{r contrast-setup}
(cm <- makeContrasts(dex=trt - untrt, levels=des))
```

<div class="note">
Many of the GSEA methods provided by limma support arbitrarily
complex experimental designs. For instance, we can consider the `cell` factor
to be a random effect (as opposed to the fixed effect we've setup above) and use
limma's `duplicateCorrelation` functionality to model this, like so:

```{r random-effects, eval=FALSE}
## This code isn't runnable now. It is included here to provide an
## example of the more exotic models that multiGSEA can support by
## virtue of leveraging limma internally.
des.re <- model.matrix(~ 0 + group, y.all$samples)
vm.re <- voom(y, des.re)
corfit <- duplicateCorrelation(vm.re, vm.re$design, block=y$samples$cell)
vm.re <- voom(y, des.re, plot=TRUE, block=y$samples$cell,
              correlation=corfit$consensus)
```

It is worth keeping in mind that that you can leverage these same moves in
limma's gene set enrichment methods as well (ie. camera, fry/(m)roast, and
romer).
</div>

### Differential Gene Expression

In this section, we first show you the straightforward analysis you would do
if you were only testing for differential gene expression.

As with any voom-based analysis, our first job is to remove lowly expressed
genes across our dataset. Here we consider a gene to be lowly
expressed if it's not expressed above 1 cpm in at least 4 experiments (because
this dataset has n=4 replication between dex treatments).

```{r gene-filter, fig.width=5, fig.height=4.5}
keep <- rowSums(cpm(y.all) >= 1) >= 4
y <- y.all[keep,,keep.lib.sizes=FALSE] %>%
  calcNormFactors %>%
  set_rownames(., .$genes$entrez_id) ## using entrez identifiers
vm <- voom(y, des, plot=TRUE)
```

Now that we have our expression (`vm`), design (`vm$design`) and contrast (`cm`)
data in hand, we can perform our differential expression analyss.

```{r dge-analysis}
fit <- lmFit(vm, vm$design) %>%
  contrasts.fit(cm) %>%
  eBayes
tt.dex <- topTable(fit, 'dex', n=Inf, sort.by='none')
```

### Gene Set Enrichment Analysis

Given that we now have all of the pieces of data required for a differential
expression analysis, performing GSEA is trivial using the `multiGSEA` wrapper
function. We simply need to now define (1) the battery of gene sets we want to
test against, and (2) the GSEA methods we want to explore.

#### Gene Sets to Test

The *multiGSEA* package provides a `GeneSetDb` class, which is a new container
(insted of `GSEABase::GeneSetCollection`) to house collections of genesets. You
can easily create a `GeneSetDb` object from a custom set of gene sets you have
on hand, but the package also provides convenience wrappers to generate 
`GeneSetDb` objects from popular gene annotation resources such as
[MSigDB][msigdb], [PANTHER][pantherdb], etc.

We'll use the `getMSigGeneSetDb` convenience function provided by the
*multiGSEA* package to simply get the hallmark geneset collection.

[msigdb]: http://software.broadinstitute.org/gsea/msigdb/
[hallmark]: http://software.broadinstitute.org/gsea/msigdb/collections.jsp#H
[c2]: http://software.broadinstitute.org/gsea/msigdb/collections.jsp#C2
[c5]: http://software.broadinstitute.org/gsea/msigdb/collections.jsp#C5
[c7]: http://software.broadinstitute.org/gsea/msigdb/collections.jsp#C7
[pantherdb]: http://pantherdb.org
[pgoslim]: http://www.pantherdb.org/panther/ontologies.jsp

```{r build-gdb, eval=!exists('gdb')}
library(multiGSEA)
gdb <- getMSigGeneSetDb(c('h', 'c2'),  'human')
```

You can view a table of the gene sets defined inside a `GeneSetDb` (`gdb`)object
via its `geneSets(gdb)` accessor:

```{r geneSets-accessor}
geneSets(gdb) %>%
  head %>%
  dplyr::select(1:4) %>%
  knitr::kable()
```

For more details on creating and manipluating `GeneSetDb` objects, please jump
the to *[The GeneSetDb Class](#the-genesetdb-class)* section.

#### Running multiGSEA

Performing multiple gene set enrichment analyses over your contrast of interest
simply requires you to provide a `GeneSetDb` object along with your data and an
enumeration of the methods you want to use in your analysis.

The call to `multiGSEA` will perform these analyses and return a
`MultiGSEAResult` object which you can then use for downstream analysis.

```{r run-multi-GSEA, eval=!exists('mg.dex'), cache=TRUE, results='hide'}
mg.dex <- multiGSEA(gdb, vm, vm$design, cm[, 'dex'],
                    methods=c('camera', 'fry', 'goseq'),
                    ## GSEA specific parameters
                    feature.max.padj=0.05, feature.min.logFC=1,
                    feature.bias=setNames(vm$genes$length, rownames(vm)),
                    inter.gene.cor=0.01)
```

We will unpack the details of the `multiGSEA` call shortly ...

#### Implicit Differential Expression

First, let's note that in addition to running a plethora of GSEA's over our data
we've also run a standard differential expression analysis. If you've passed
a `matrix`, `ExpressionSet` or `EList` into `multiGSEA`, a *limma*-based
`lmFit  %>% (eBayes|treat) %>% (topTable|topTreat)` pipeline was run. If a
`DGEList` was passed, then `multiGSEA` utilizes the *edgeR*-based
`glmQLFit %>% (glmQLFTest | glmTreat) %>% topTags` pipeline.

Either way, the results of the internally run differential expression analysis
is accessible via a call to `logFC` function on the `MultiGSEAResult` object:

```{r logFC-results, results='asis'}
lfc <- logFC(mg.dex)
lfc %>%
  dplyr::select(symbol, entrez_id, logFC, CI.L, CI.R, t, pval, padj) %>%
  dplyr::slice(1:5) %>%
  knitr::kable(digits=3)
```

We can confirm that the statistics generated internally in multiGSEA mimic our
explicit analysis above by verifying that the t-statistics generated by both
approaches are identical.

```{r compare-dge-t-stats}
comp <- tt.dex %>% 
  dplyr::select(entrez_id, symbol, logFC, t) %>%
  inner_join(lfc, by=c('entrez_id'), suffix=c('.tt', '.mg'))
all.equal(comp$t.tt, comp$t.mg)
```

```{r plot-dge-t-stats, fig.width=4.5, fig.height=4}
ggplot(comp, aes(t.tt, t.mg)) +
  geom_point(alpha=0.7) +
  geom_abline(slope=1, intercept=0, color='red') +
  xlab("t-statistics (external voom piepline)") +
  ylab("t-statistics (internal multiGSEA)") +
  ggtitle("t-statistics comparison")  
```

The internally performed differential expression analysis within the `mulitGSEA`
call can be customized almost as extensively as an explicitly performed analysis
that you would run using limma or edgeR by sending more paramaters through
`multiGSEA`'s `...` argument.

See the
*[Custom Differential Expression](#custom-differential-expression)*
section further in the vignette as well as the help available in
`?calculateIndividualLogFC` (which is called inside the `multiGSEA` function)
for more information.

#### Explicit GSEA

We also have the results of all the GSEA analyses that we specified to our
multiGSEA call via the `methods` parameter.

```{r mg-res}
mg.dex
```

The table above enumerates the different GSEA methods run over each geneset
collection in the rows. The columns enumerate the number of genesets that the
collection has in total (`geneset_count`), and how many were found significant
at a given FDR (here it's FDR <= 30%).

### Exploring Results

GSEA results can be examined interactively via the command line, or via a shiny
app. You can use the `resultNames` function to find out what GSEA methods were
run, and therefore available to you, within the the `MultiGSEAResult` object:

```{r resultnames}
resultNames(mg.dex)
```

Note that when running a "goseq" analysis, multiGSEA will (by default) run
it three different ways. By running an enrichment analysis on all differentially
expressed genes, then separately for only genes that go up in your contrast,
and a third time for only the genes that go down.

The individual gene set statistics generated by each method are available via
the `result` function (or several can be returned with `results`):

```{r goseq-res}
cam.res <- result(mg.dex, 'camera')
cam.go.res <- results(mg.dex, c('camera', 'goseq.up'))
```

You can also list the members of a geneset and their individual differential
expression statistics for the contrast under test using the `geneSet` function.

```{r geneset-result}
gs <- geneSet(mg.dex, 'c2', 'STEGER_ADIPOGENESIS_UP')
gs %>%
  dplyr::select(symbol, entrez_id, logFC, pval, padj) %>%
  head %>%
  knitr::kable(digits=3)
```

<div class="note">
The results provided in the table generated from a call to `geneSet` are
independant of GSEA method.
</div>

#### Plotting

*multiGSEA* provides a number of interactive plotting facilities to explore the
enrichment of a single geneset under the given contrast. In the boxplots and
density plots shown below, the log fold changes (logFCs) (or t-statistics) for
all genes under the contrast are visualized in the "background" set, and these
same values are shown for the desired geneset under the "geneset" group.

**Boxplot**

```{r iplot-tnfa-box, fig.width=4, fig.height=3}
iplot(mg.dex, 'c2', 'STEGER_ADIPOGENESIS_UP', type='boxplot', value='logFC')
```

**Density**

```{r iplot-tnfa-density, fig.width=4, fig.height=3}
iplot(mg.dex, 'c2', 'STEGER_ADIPOGENESIS_UP', type='density', value='logFC')
```

#### Interactive Exploration

A shiny application is included in the `inst/shiny` folder of this package.
This shiny application can be launched via tha `explore` function so that
the analyst can explore the results stored in a MultiGSEAResult "in real time."
The application can also be deployed to a shiny server, and users can upload
`MultiGSEAResult` objects in an rds object (via `saveRDS`).

Run `explore(mg.dex)` to see it in action.

# The GeneSetDb Class

The GeneSetDb class is a new container to store collections of genesets which
provides different types of functionality than is found in
GSEABase::GeneSetCollection objects.

We can, for instance, identify all genesets that have a specific gene(s) like
so:

```{r subset-gdb}
gdb <- geneSetDb(mg.dex)
gdb.sub <- subsetByFeatures(gdb, c('2040', '10062'))
```

<!--
  * Constructing custom `GeneSetDb`s:
      * From a `list` of `lists` 
      * From a `data.frame`
  * `gdb <- conform(gdb, vm)`
      * Identifies the features you've measured in the experiment `vm`
      * Appropriately sets the `universe` of tests that require it
      * It's "non-destructive", you can reconform the same `GeneSetDb` to
        another dataset without loss of information (geneset membership)
      * Use of `featureIdMap`
-->

## Building a GeneSetDb

The multiGSEA package provides conveneince funcitons to fetch genesets from many
sources and convert them into a GeneSetDb object. The two most useful sources
may be:

* [MSigDB][msigdb] via `getMSigGeneSetDb(...)`
* [PANTHER][pantherdb] (pathways and GOSLIM) via `getPantherGeneSetDb()`

The `GeneSetDb` constructor also can take a number of different inputs and
convert them into a GeneSetDb, such as:

1. A data.frame of geneset membership. This requires `collection`, `name`, and
  `featureId` columns.
2. A named list of gene identifer vectors that represent genesets for a
   single collection
3. A named list of (2)-like lists. The top level names are the names of the
   different collecitons, and each sublist represents the genesets in that
   collection.
  
See the help and examples in ?GeneSeDb for more information.

## Active vs Inactive Gene Sets

After a `GeneSeDb` is conformed to a target expression object `E`,
individual gene sets become "active" or "inactive" depending on the number
of features per gene set are found in `E`. For instance, if `E`
does not have measurements for at least `conform(, min.gs.size)`
features for a given gene set, that gene set will be flagged as inactive.

## Mapping of gene set featureIds to target expression containers

It may be that the IDs used in a gene set collection are different from the
ones used as the rownames of your expression cotainer. For instance, the IDs
used for a given gene set collection in the \code{GeneSetDb} might be
Ensembl gene identifiers, but the rownames of the expression object migth
be Entrez ID. This is where the \code{mapping} parameter becomes useful.

NOTE: this functionality is under construction. It is not fully complete
and the documentation therefore can not be written. I do want to leave a
section here for future expansion, though.

# Costumizing Analyses

The `...` in `multiGSEA(gdb, x, design, coef, methods, ...)` are you friend.

## Custom Differential Expression

The internal differential expression pipeline, exported via the
`calculateIndividualLogFC` function allows the end user to configure an
"arbirarily comples" differential expression analysis using either edgeR's
quasilikelihood framwork (if the input is a DGEList) or a direct limma
analysis (with a pre-voomed EList, expression matrix, or whatever).

User's should refer to the `?calculateIndividualLogFC` help page to see 
which parameters are exposed for a differential expression analysis and
configure them accordingly. When calling `multiGSEA` these parameters
can also be specified and will be passed down to the internal
`calculateIndividualLogFC` call.

## Custom GSEA


# Extending multiGSEA

## Adding new GSEA methods

There are three funtions you need to provide in order to add a new method `XXX`
to the multi `multiGSEA` package:

  * `validate.inputs.XXX`
  * `do.XXX`
  * `summarize.XXX`

### Validate Inputs

In general, it seems that GSEA methods require "just" a vector of statistics
that enrichment can be run over, or a full representation of an experiment and
a specific contrast to test.

If your method simply requires a vector of statistics, like
\code{limma::geneSetTest}, then alias your `validate.inputs.XXX` function to
`multiGSEA::.validate.inputs.logFC.only`. If the method is of the latter
variety, like \code{limma::camera} then alias your `validate.inputs.XXX` method
to `multiGSEA:::..validate.inputs.full.design`, ie:

    validate.inputs.my_new_method <- multiGSEA:::..validate.inputs.full.design

### Do the analysis

The `do.XXX` method just support the following signature:

    do.my_new_method(gsd, x, design, contrast, ...)

If your method only requires a vector of statistics, `design` and `contrast`
your method can simply ignored `design` and `contrast` (or, you can let your
`...` consume them).

### Summary of results

The `sumamrize.XXX` method should return a data.table that (minimally) had
a `collection` and `id` column followd by relevant information/statistics
produced by your method for each geneset that was tested.

# Scratch

## Background reading for Gene Set Enrichment Analyses


Overview of The Broad's GSEA:
http://baderlab.org/CancerStemCellProject/VeroniqueVoisin/AdditionalResources/GSEA


## Himes airway paper

Quotes from the Himes et al. paper:

* Differential Gene Expression:

    > Overall, 316 genes were significantly differentially expressed after correcting for false discovery rate by the Benjamini-Hochberg [23] approach [Figure 1A, Table S3]. Table 1 contains the genes with Q-value <1E-10

* GSEA:

    > Gene set enrichment analysis using the NIH DAVID tool [28] identified various Gene Ontology and other annotation categories that were overrepresented by the 316 genes. The top six gene functional annotation clusters (enrichment scores >3) had terms related to: glycoprotein/extracellular matrix, vasculature development, circulatory system process, response to nutrients, thrombospondin type-1, and response to hormone stimulus terms [Table S4]. Other clusters among the 19 with enrichment scores >1.5 that may be relevant to lung disease included lung development, regulation of cell migration, and extracellular matrix organization.
