---
title: "Performing gene set enrichment analyses with multiGSEA"
author: "Steve Lianoglou"
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('multiGSEA')`"
abstract: >
  The multiGSEA package facilitates the execution and comparison of several
  GSEA approaches for a single experimental contrast and provides tools
  to explore these results from within the user's workspace or via a shiny
  application.
  multiGSEA package version: `r packageVersion("multiGSEA")`
output:
  rmarkdown::html_document:
    self_contained: yes
    highlight: pygments
    toc: true
    fig_width: 5
vignette: >
  %\VignetteIndexEntry{Performing gene set enrichment analyses with multiGSEA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding[utf8]{inputenc}
---

```{r init, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
library(multiGSEA)
knitr::opts_chunk$set(
  echo=TRUE, warning=FALSE, message=FALSE, error=FALSE) #, dpi=150)
```

# Overview

The `multiGSEA` package was built to facilitate the use of gene sets in the
analysis of high throughput genomics data (primarily RNA-seq). It does so
by providing these top-line functionalities:

* The `multiGESA` function is a wrapper that orchestrates the execution of any
  number of user-specified gene set enrichment analyses (GSEA) over a particular
  experimental contrast of interest. This will create a `MultiGSEAResult`
  object which stores the results of each GSEA method internally, allowing
  for easy query and retrieval.
* The `explore` function launches a shiny application that facilitates the
  interactive exploration of these GSEA results. This application lives in
  `inst/shiny` and can also be deployed to a shiny server and can be initialized
  by uploading a serialized MultiGSEAResult *.rds file.
* The `scoreSingleSamples` function is a wrapper that enables the user to
  generate single sample gene set scores using a variety of different
  single sample gene set scoring methods.
* The `GeneSetDb` class: a new class to store collections of gene sets. This
  provides a bit more functionality than the base `GeneSetCollection` class
  defined in the [GSEABase](http://bioconductor.org/packages/GSEABase).

The initial GSEA methods that multiGSEA wrapped were the ones provided by limma
and edgeR. As such, many analyses using multiGSEA expect you to re-use the same
data objects used for differential expression analysis, namely:

* Expression data (an `EList`, `DGEList`, or expression matrix)
* A design matrix
* A contrast vector/matrix (if your design and comparison require it)

Other methods only require the user to provide a vector of statistics that
represent some differential expression statistic per gene, and the GSEA is
performed by analyzing the ranks of genes within this vector.

All GSEA methods require the use of a `GeneSetDb`.

Currently supported gene set enrichment methods include:

```{r multiGSEA-methods, results='asis', echo=FALSE}
meths <- setdiff(multiGSEA.methods(), c('logFC', 'svdGeneSetTest'))
cat(sprintf('* %s\n', meths))
```

When using these methods in analyses that lead to publication, please cite
the original papers that developed these methods and cite multiGSEA when its
functionality assisted in your interpretation and analysis.

The multiGSEA package provides a small example expression dataset extracted from
the TCGA BRCA dataset, which is available via the `exampleExpressionSet`
function. In this vignette we will explore differential expression and gene
set enrichment analysis by examining differences between basal and her2 PAM50
subtypes.

# Standard Workflow

Let's begin by setting up our work environment for exploratory analysis using
the multiGSEA package.

```{r init-env}
library(multiGSEA)
library(magrittr)
library(dplyr)
library(ggplot2)
theme_set(theme_bw())
options(multiGSEA.df.return='data.frame')
```

<div class="note">
Internally, mulitGSEA leverages the
[`data.table`](https://CRAN.R-project.org/package=data.table) package for fast
indexing and manipulation over data.frames. There are many functions that return
data.tables back to the end user, but some users may not want to work with those.
The `multiGSEA.df.return` option can be set to `data.frame` (as above) so that
all data.frame-like objects returned by this package are actually `data.frames`,
and not `data.tables`.
</div>

## Data Setup

multiGSEA is most straightforward to use when our data objects and analysis are
performed with either the edgeR or voom/limma pipelines and when we use
Entrez IDs for gene gene identifiers.

The `exampleExpressionSet` function gives us just such an object. We call it
below in a manner that gives us an object that allows us to explore expression
differences between different subtypes of breast cancer.

```{r data-setup, eval=!exists('y.all'), results='hide'}
vm <- exampleExpressionSet(dataset='tumor-subtype', do.voom=TRUE)
```

Below you'll find the `$targets` data.frame of the voomed `EList`

```{r}
vm$targets[, c('Patient_ID', 'Cancer_Status', 'PAM50subtype')]
```

## Data Analysis

We will identify the genes and genesets that are differentially expressed
between the basal and her2 subtypes. The `vm` object has already been `voom`d
using this design:

```{r}
vm$design
```

We can test for differences between basla and her2 subtypes using the following
contrast:

```{r contrast-setup}
(cm <- makeContrasts(BvH=Basal - Her2, levels=vm$design))
```

### Differential Gene Expression

In this section, we first show you the straightforward analysis you would do
if you were only testing for differential gene expression.

With the data we have at hand, you would simply do the following:

```{r dge-analysis}
fit <- lmFit(vm, vm$design) %>%
  contrasts.fit(cm) %>%
  eBayes
tt <- topTable(fit, 'BvH', n=Inf, sort.by='none')
```

### Gene Set Enrichment Analysis

Given that we now have all of the pieces of data required for a differential
expression analysis, performing GSEA is trivial using the `multiGSEA` wrapper
function. We simply need to now define (1) the battery of gene sets we want to
test against, and (2) the GSEA methods we want to explore.

#### Gene Sets to Test

The *multiGSEA* package provides a `GeneSetDb` class, which is used instead of
a `GSEABase::GeneSetCollection` to house collections of genesets. You
can easily create a `GeneSetDb` object from a custom set of gene sets you have
on hand, but the package also provides convenience wrappers to generate 
`GeneSetDb` objects from popular gene annotation resources such as
[MSigDB][msigdb], [PANTHER][pantherdb], etc.

We'll use the `getMSigGeneSetDb` convenience function provided by the
*multiGSEA* package to load the [hallmark][hallmark] (`"h"`) and
[c2 (curated)][c2] (`"c2"`) gene set collections from [MSigDB][msigdb].

[msigdb]: http://software.broadinstitute.org/gsea/msigdb/
[hallmark]: http://www.cell.com/cell-systems/abstract/S2405-4712(15)00218-5
[isigdb]: http://www.cell.com/immunity/abstract/S1074-7613(15)00532-4
[c2]: http://software.broadinstitute.org/gsea/msigdb/collections.jsp#C2
[c5]: http://software.broadinstitute.org/gsea/msigdb/collections.jsp#C5
[c7]: http://software.broadinstitute.org/gsea/msigdb/collections.jsp#C7
[pantherdb]: http://pantherdb.org
[pgoslim]: http://www.pantherdb.org/panther/ontologies.jsp

```{r build-gdb, eval=!exists('gdb')}
gdb <- getMSigGeneSetDb(c('h', 'c2'), 'human')
```

You can view a table of the gene sets defined inside a `GeneSetDb` (`gdb`)object
via its `geneSets(gdb)` accessor:

```{r geneSets-accessor}
geneSets(gdb) %>%
  head %>%
  select(1:4)
```

For more details on creating and manipulating `GeneSetDb` objects, please jump
the to *[The GeneSetDb Class](#the-genesetdb-class)* section.

#### Running multiGSEA

Performing multiple gene set enrichment analyses over your contrast of interest
simply requires you to provide a `GeneSetDb` object along with your data and an
enumeration of the methods you want to use in your analysis.

The call to `multiGSEA` will perform these analyses and return a
`MultiGSEAResult` object which you can then use for downstream analysis.

```{r run-multi-GSEA, eval=!exists('mg'), results='hide'}
mg <- multiGSEA(
  gdb, vm, vm$design, cm[, 'BvH'],
  methods=c('camera', 'fry', 'goseq'),
  ## these parameters define which genes are differentially expressed
  feature.max.padj=0.05, feature.min.logFC=1,
  ## for camera:
  inter.gene.cor=0.01,
  ## for goseq (the *.max.padj and *.min.logFC parameters also affect goseq)
  feature.bias=setNames(vm$genes$size, rownames(vm)))
```

We will unpack the details of the `multiGSEA` call shortly ...

#### Implicit Differential Expression

First, let's note that in addition to running a plethora of GSEA's over our data
we've also run a standard differential expression analysis. If you've passed
a `matrix`, `ExpressionSet` or `EList` into `multiGSEA`, a *limma*-based
`lmFit  %>% (eBayes|treat) %>% (topTable|topTreat)` pipeline was run. If a
`DGEList` was passed, then `multiGSEA` utilizes the *edgeR*-based
`glmQLFit %>% (glmQLFTest | glmTreat) %>% topTags` pipeline.

Either way, the results of the internally run differential expression analysis
is accessible via a call to `logFC` function on the `MultiGSEAResult` object:

```{r logFC-results}
lfc <- logFC(mg)
lfc %>%
  select(symbol, entrez_id, logFC, t, pval, padj) %>%
  head
```

We can confirm that the statistics generated internally in multiGSEA mimic our
explicit analysis above by verifying that the t-statistics generated by both
approaches are identical.

```{r compare-dge-t-stats}
comp <- tt %>% 
  select(entrez_id, logFC, t, pval=P.Value, padj=adj.P.Val) %>%
  inner_join(lfc, by='entrez_id', suffix=c('.tt', '.mg'))
all.equal(comp$t.tt, comp$t.mg)
```

The internally performed differential expression analysis within the `mulitGSEA`
call can be customized almost as extensively as an explicitly performed analysis
that you would run using limma or edgeR by sending more parameters through
`multiGSEA`'s `...` argument.

See the
*[Custom Differential Expression](#custom-differential-expression)*
section further in the vignette as well as the help available in
`?calculateIndividualLogFC` (which is called inside the `multiGSEA` function)
for more information.

#### Explicit GSEA

We also have the results of all the GSEA analyses that we specified to our
multiGSEA call via the `methods` parameter.

```{r mg-res}
mg
```

The table above enumerates the different GSEA methods run over each geneset
collection in the rows. The columns enumerate the number of genesets that the
collection has in total (`geneset_count`), and how many were found significant
at a given FDR, which is set to 20% by default. The `show` command for the
`MultiGSEAResult` object simply calls the `tabulateResults` function, which
you can call directly with the value of `max.p` that you might find more
appropriate.

### Exploring Results

GSEA results can be examined interactively via the command line, or via a shiny
application. You can use the `resultNames` function to find out what GSEA
methods were run, and therefore available to you, within the the
`MultiGSEAResult` object:

```{r resultnames}
resultNames(mg)
```

Note that when running a "goseq" analysis, multiGSEA will (by default) run
it three different ways. By running an enrichment analysis on all differentially
expressed genes, then separately for only genes that go up in your contrast,
and a third time for only the genes that go down.

The individual gene set statistics generated by each method are available via
the `result` function (or several can be returned with `results`):

```{r goseq-res}
cam.res <- result(mg, 'camera')
cam.go.res <- results(mg, c('camera', 'goseq.up'))
```

You can identify genesets with the strongest enrichment by filtering and sorting
against the appropriate columns. We can, for instance, identify which hallmark
gene sets show the strongest enrichment as follows:

```{r}
cam.res %>% 
  filter(padj < 0.1, collection == 'h') %>% 
  arrange(desc(mean.logFC)) %>% 
  select(name, n, mean.logFC, padj) %>% 
  head
```

You can also list the members of a geneset and their individual differential
expression statistics for the contrast under test using the `geneSet` function.

```{r geneset-result}
geneSet(mg, 'h', 'HALLMARK_WNT_BETA_CATENIN_SIGNALING') %>% 
  select(symbol, entrez_id, logFC, pval, padj) %>%
  head
```

<div class="note">
The results provided in the table generated from a call to `geneSet` are
independant of GSEA method. The statistics appended to the gene set members
are simply the ones generated from a differential expression analysis.
</div>

#### Plotting

*multiGSEA* provides a number of interactive plotting facilities to explore the
enrichment of a single geneset under the given contrast. In the boxplots and
density plots shown below, the log fold changes (logFCs) (or t-statistics) for
all genes under the contrast are visualized in the "background" set, and these
same values are shown for the desired geneset under the "geneset" group.

The logFC (or t-statistics) of the genes in the gene set are plotted as points,
which allow you to hover to identify the identity of the genes that land in
the regions of the distributions you care about.

**Boxplot**

```{r iplot-tnfa-box, fig.asp=1}
iplot(mg, 'h', 'HALLMARK_WNT_BETA_CATENIN_SIGNALING',
      type='boxplot', value='logFC')
```

**Density**

```{r iplot-tnfa-density, fig.asp=1}
iplot(mg, 'h', 'HALLMARK_WNT_BETA_CATENIN_SIGNALING',
      type='density', value='logFC')
```

#### Interactive Exploration

A shiny application is included in the `inst/shiny` folder of this package.
This shiny application can be launched via tha `explore` function so that
the analyst can explore the results stored in a MultiGSEAResult "in real time."
The same application can also be deployed to a shiny server, which allows users
to upload serialized `MultiGSEAResult` objects as rds objects (via `saveRDS`).

Run `explore(mg)` to see this application in action.

A separate vignette will be written that gives a tour of this application in
more detail.

# The GeneSetDb Class

The GeneSetDb class is a new container to store collections of genesets which
provides different types of functionality than is found in
GSEABase::GeneSetCollection objects.

We can, for instance, identify all 84 gene sets that have genes "10014" and
"1454" (entrez ids) as members (HDAC5 and CSNK1E, respectively).

```{r subset-gdb}
gdb <- geneSetDb(mg)
gdb.sub <- subsetByFeatures(gdb, c('10014', '1454'))
nrow(gdb); nrow(gdb.sub)
```

The GeneSetDb object uses the `data.table` package internally for fast lookup.
The code will be optimized in the future to be even more performant. Internally
the collection of gene set information is minimally stored as a three-column
`data.table` in "long form", which has the following columns:

* collection
* name
* featureId

More columns can be added to the internal `data.table` (a "symbol" column, for
instance), but those are the only three you need. To see what we are talking
about, exactly, you can call the `as.data.frame` function on a `GeneSetDb`
object:

```{r}
as.data.frame(gdb)[c(1:5, 201:205),]
```


The `(collection,name)` tuple is the primary key of a gene set. The `featureId`
column stores gene identifiers. For the time being, it will be most natural
for these featureIds to simply be entrez ids, as many of the annotation
databases use entrez gene identifiers themselves.

## Building a GeneSetDb

The multiGSEA package provides convenience funcitons to fetch genesets from many
sources and convert them into a GeneSetDb object. The two most useful sources
may be:

* [MSigDB][msigdb] via `getMSigGeneSetDb(...)`
* [PANTHER][pantherdb] (pathways and GOSLIM) via `getPantherGeneSetDb()`

You can create a custom `GeneSetDb` via the `GeneSetDb` constructor, which
accpets the following inputs:

1. A data.frame of geneset membership. This requires `collection`, `name`, and
   `featureId` columns. Reference the output of `as.data.frame(gdb)` shown
   above.
2. A named list of gene identifier vectors that represent genesets for a
   single collection
3. A named list of (2)-like lists. The top level names are the names of the
   different collections, and each sublist represents the genesets in that
   collection.

Two `GeneSetDb` objects can be combined using the `append` function. For now
it is your responsibility to ensure that the two `GeneSetDb` objedts are
"reasonably conformable", ie. they use the same types of gene identifiers, and
are referencing the same species, etc.

```{r, eval=FALSE}
msigdb <- getMSigGeneSetDb('h', 'human')
goslimdb <- getPantherGeneSetDb('goslim', 'human')
gdb.uber <- append(msigdb, goslimdb)
```

See the help and examples in `?GeneSetDb` for more information.

## Active vs Inactive Gene Sets

A `GeneSetDb` is used to hold "the universe" of genes that belong to different
gene sets across different collections. Depending on the assay performed to 
measure these genes, the set of genes you observe in your study will likely
be a subset of the genes in the `GeneSetDb`. As such, prior to using a
`GeneSetDb` for GSEA, it must be "conformed" to a target object that will be
used for the input to the GESA (either a matrix of expression, or a pre ranked
vector of statistics). This step will index into the target expression object
and identify which rows of the object correspond to which genes in the
`GeneSetDb`.

"Conformation" happens automatically within the `multiGSEA` call, but we call it
explicitly below to outline its functionality. The command below conforms
the `GeneSetDb` to our target "voomed" `EList`, and deactivates gene sets
(i.e. removes them from downstream GSEA) that have less than 10 or more than 100
genes that were found in `vm`:

```{r}
gdbc <- conform(gdb, vm, min.gs.size=10, max.gs.size=100)
head(geneSets(gdbc, active.only=FALSE))
```

We can see that, only 23 of the 26 genes in the
`(c2,ABBUD_LIF_SIGNALING_1_DN)` were found in the rows of `vm`, and the `(c2,ABBUD_LIF_SIGNALING_2_DN)` was "deactivated." Deactivated
(`active == FALSE`) gene sets will be ignored during downstream analyses. This
gene set was deactivated  because it only has five "conformed" genes, but the
minimum geneset size we wanted to consider (`min.gs.size`) was set to ten in
our call to `conform`.

## Accessing members of a gene set

The `geneSet` and `featureIds` functions allow the user to identify the genes
found in a gene set. Both of these functions take an `active.only` argument,
which is `TRUE` by default. This specifies that only the genes that have been
successfully conformed to a gene set should be the ones that are returned.

For instance, we can identify which genes belong to the
`(c2,ABBUD_LIF_SIGNALING_1_DN)`, and which three were not found in `vm` like so:

```{r}
missed <- setdiff(
  featureIds(gdbc, 'c2', 'ABBUD_LIF_SIGNALING_1_DN', active.only=FALSE),
  featureIds(gdbc, 'c2', 'ABBUD_LIF_SIGNALING_1_DN', active.only=TRUE))
missed
```

or we can use the `geneSet` function to return a `data.frame` of these results:

```{r}
gdbc %>% 
  geneSet('c2', 'ABBUD_LIF_SIGNALING_1_DN', active.only=FALSE) %>%
  subset(featureId %in% missed)
```

## Mapping of gene set featureIds to target expression containers

It may be that the IDs used in a gene set collection are different from the
ones used as the rownames of your expression container. For instance, the IDs
used for a given gene set collection in the `GeneSetDb` might be
Ensembl gene identifiers, but the rownames of the expression object might
be Entrez ID. This is where the `mapping` parameter becomes useful.

The `GeneSetDb` class has a concept of an internal `featureIdMap` to accommodate
these scenarios, which will be further developed in future versions. For the
time being, the user is required to manually map the feautreIds in their
expression matrix to be concordant with the ones found in the `GeneSetDb`.

# Customizing Analyses

The internal differential expression analysis as well the gene set enrichment
analyses can be customized by passing parameters through the `...` in the
`multiGSEA` function.

## Custom Differential Expression

The internal differential expression pipeline, exported via the
`calculateIndividualLogFC` function allows the end user to configure an
"arbitrarily complex" differential expression analysis using either edgeR's
quasilikelihood framework (if the input is a DGEList) or a direct limma
analysis (with a pre-voomed EList, expression matrix, or whatever).

User's should refer to the `?calculateIndividualLogFC` help page to see 
which parameters are exposed for a differential expression analysis and
configure them accordingly. When calling `multiGSEA` use these same parameters
in the call and they will be provided to `calculateIndividualLogFC`.

For instance, if you wanted to use limma's "treat" functionality to specify
a minimal log fold change threshold for statistical significance, you would do
so as follows:

```{r, eval=FALSE}
mg <- multiGSEA(gdb, vm, vm$design, cm[, 'BvH'],
                methods=c('goseq'),
                treat.lfc=log2(1.5))
```

Using the internal `treat` functionality would really only affect enrichment
tests that first threshold the genes in your experiment as "significant" or not,
like `goseq` and not tests like `camera`.

## Custom GSEA

The GSEA methods that are wrapped by `multiGSEA` all take the same parameters
that are defined by their implementation. Simply pass these parameters down
via the `...` in the `multiGSEA` call.

For instance, you can read `?camera` to find that the `camera` method accepts an `inter.gene.cor` parameter, and `?roast` will tell you that you can specify
the number of rotations used via the `nrot` parameter.

```{r, eval=FALSE}
mgx <- multiGSEA(gdb, vm, vm$design, cm[, 'BvH'],
                 methods=c('camera', 'roast'),
                 inter.gene.cor=0.04, nrot=500)
```

# Single Sample Gene Set Scoring

Gene sets can also be used outside of these "formal testing frameworks" in the
"singe sample scoring" scenario. The `scoreSingleSamples` method is a wrapper
function to these types of methods the same way that the `multiGSEA` function
wraps formal gene set enrichment analysis methods.

The `scoreSingleSamples` function will return a long data.frame that has scores
on a per sample, per gene set, per method basis. For instance, we can get ssGSEA
(from the GSVA package) and mean Z scores for the genesets per sample like so:

```{r}
h <- getMSigGeneSetDb('h', 'human')
scores <- scoreSingleSamples(h, vm, methods=c('ssgsea', 'zscore'),
                             ssgsea.norm=TRUE)
```

And we can compare the results of ssgsea vs zscores by manipulating the result
into a wide data.frame and plotting these scores against each other:

```{r, message=FALSE, fig.asp=0.85}
suppressPackageStartupMessages(library(tidyr))
sw <- spread(scores, method, score)
sw.some <- filter(sw, name %in% c('HALLMARK_ADIPOGENESIS', 'HALLMARK_APOPTOSIS'))
ggplot(sw.some, aes(zscore, ssgsea)) +
  geom_point(aes(color=name))
```

