---
title: "multiGSEA workflow in TCGA Breast"
author: Steve Lianoglou
date: "`r format(Sys.time())`"
output:
  html_document:
    self_contained: false
    toc: true
    toc_depth: 3
    lib_dir: libs
knit: rmd.plugins::report.render
vignette: |
  %\VignetteIndexEntry{multiGSEA workflow in TCGA Breast}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r setup,echo=FALSE,results="hide",warning=FALSE,message=FALSE,error=FALSE}
library(Biobase)
library(edgeR)
library(multiGSEA)
library(rmd.plugins)
library(knitr)
library(magrittr)
## Set defaults to hide errors and warnings, show output, and create highres
## pngs
report.init(echo=TRUE, results='markup')
```

# Overview

This document is meant to show "a typical" differential expression and gene
set enrichment analysis for a project using the `rmd.plugins` and `multiGSEA`
packages.

To compile this document, you simply have to invoke the `report.render` function
from the `rmd.plugins` package like so:

```{r compile-example, eval=FALSE}
## Note that specifying `gen.dir` is optional
report.render("vignettes/tcga-breast-gsea.Rmd", gen.dir='_rmd_plugins_gen_brca')
```

<div class="tip">
This document defines `knit: rmd.plugins::report.render` in the
[YAML frontmatter][yamlknit], and therefore can be compiled directly from within
RStudio by simply calling the "Knit" command (the "Knit" button in the tool bar,
or the ⇧⌘K key  combination)
</div>

[yamlknit]: http://www.r-bloggers.com/one-weird-trick-to-compile-multipartite-dynamic-documents-with-rmarkdown/

Let's define a few variables that will be used to customize the output:

```{r init-vars, cache=TRUE}
## The GSEA methods we want to run
gsea.methods <- c('camera', 'geneSetTest', 'hyperGeometricTest', 'roast')

## The ExpessionSet we will be analyzing. multiGSEA provides a convenience
## function to provide a subset of samples from the TCGA/BRCA indication
es <- exampleExpressionSet(do.voom=FALSE)

## Currently multiGSEA only supports RNA-seq data if it has been voomed, or if
## is passed in as a DGEList
es.design <- local({
  d <- model.matrix(~ Cancer_Status, pData(es))
  colnames(d) <- sub('Cancer_Status', '', colnames(d))
  d
})

y <- DGEList(exprs(es), group=pData(es)$Cancer_Status)
y <- calcNormFactors(y)
y <- estimateDisp(y, es.design)

vm <- voom(y, es.design, plot=FALSE)
```

We will want to also define the universe of genesets that we want to test. The
`multiGSEA` package provides helper functions that return the `MSigDB`
collections for mouse our human. Some collections of interest may be:

  * "h": [hallmark gene set][hsets]
  * "c2": [curated gene sets][c2sets]
  * "c7": [immunologic signatures][c7sets]

[hsets]: http://www.broadinstitute.org/gsea/msigdb/genesets.jsp?collection=H
[c2sets]: http://www.broadinstitute.org/gsea/msigdb/genesets.jsp?collection=C2
[c7sets]: http://www.broadinstitute.org/gsea/msigdb/genesets.jsp?collection=C7

So let's fetch them:

```{r genesetdb-init, results='asis', cache=TRUE}
gsd.all <- getMSigDBset(c('h', 'c2', 'c7'), 'human')
tally.all <- table(geneSets(gsd.all)$collection) %>%
  as.data.frame %>%
  setNames(c('collection', 'count'))
kable(tally.all, table.attr='style="width: 250px"')
```


There are `r nrow(geneSets(gsd.all))` gene sets defined among the `c2` and `c7`
gene set collections:

We could use these for the downstream analysis, but to
save time on computation, we will use all of the hallmark genes, and a
predefined subset of c2 and c7.

```{r genesetdb-ex, results='asis'}
gsd <- append(getMSigDBset('h'), exampleGeneSetDb())
tally <- table(geneSets(gsd)$collection) %>%
  as.data.frame %>%
  setNames(c('collection', 'count'))
kable(tally)
```

# Tumor vs. Normal

Vanilla run of multiGSEA

```{r tn-multiGSEA, cache=TRUE, dependson="init-vars"}
mg <- multiGSEA(gsd, vm, vm$design, contrast=2, methods=gsea.methods)
```

To display the results, we simply have to drop the output of the `rmd.multiGSEA`
call into a `results='asis'` code chunk.

# Exploring GSEA results

## Command Line Exploration

The results of a call to `multiGSEA` is a `MultiGSEAResult` object, which can be
interrogated at the command line. Some functions that are helpful here include:

  * `result[s](mg)`: Shows the pvals & statistics of all (or a subset) of the
    methods run over each geneset.
  * `resultNames(mg)`: Lists the results available to display
  * `tabulateResults(mg)`: Tabulates number of genesets for each collection
    that were found to be significant by each method

<div class="important">
Please refer to the help documentation for the functions above for more
information. While printing out a report from a multiGSEA run (as shown below)
is handy for downstream consumers of the analysis, interrogating these results
manually from the command line is imoprtant for the analysist.
</div>

### Plotting GeneSet results

You might be curious what the distribution of log fold changes are for the
genes in an arbitrary geneset vs the log fold changes across the entire
experiments. You can view these results in two ways:

  * A `density` plot
  * Using limma's `barcode` plot

#### Density plots

The default `plot` method for a `MultiGSEAResult` produces a density plot,
like so:

```{r mg-density-plot, eval=TRUE}
plot(mg, 'c7', 'GSE339_EX_VIVO_VS_IN_CULTURE_CD8POS_DC_UP')
```

The points at the bottom of the plot indicate the log fold change for each of
the individual items in the geneset. Points outlined in black are ones that
are significantly differentially expressed given the `feature.min.logFC` and
`feature.max.padj` criteria that were passed into the `multiGSEA` call.

#### Barcode plots

limma provides a very handy `barcodeplot` function that show Broad-like GSEA
plots (but even more informative). Setting `type="barcode"` in the call to
`plot(mg, type='barcode')` draws these plots. Currently the code that does this
has been lifted straight from limma, but the plan is to extend its functionality
in the future.

```{r mg-barcode-plot}
plot(mg, 'c7', 'GSE339_EX_VIVO_VS_IN_CULTURE_CD8POS_DC_UP', type='barcode')
```

## HTML Reports

The `rmd.plugins` package defines an `rmd.multiGSEA` function which wraps a
`MultiGSEAResult` into a browseable widget for an rmarkdown document. It
provides several parameters that can be used to customize which (and how many)
results are shown.

### Report all significant results

By default, only results with an adjusted p value of &le; 0.30 are reported
for each method. Use the `max.p` and `p.col` arguments to tweak what should be
considered significant.

```{r show-multiGSEA-tn, results='asis'}
rmd.multiGSEA('tumor-vs-normal-gsea', mg)
```

We can tweak several parameters when we report a `multiGSEA` result to customize
it.

Perhaps it's useful to plot the t-statistic on the x-axis as opposed to the
(default) log fold change:

```{r show-multiGSEA-tn-tstat, results='asis'}
rmd.multiGSEA('tumor-vs-normal-gsea-tstat', mg, plot.stat='t')
```

### Report top.n results

Maybe you want to show the top.n genesets per method irrespective of their
FDR. In this case do not pass in a value for `max.p` but rather set `top.n`
to a value of your choosing.

In the reports below, the top 10 results *per GSEA method* are reported.

```{r show-multiGSEA-top10, results='asis'}
rmd.multiGSEA('tumor-vs-normal-gsea-top10', mg, top.n=10)
```

### Report maximum of top.n significant reuslts

Maybe you want to show the `top.n` genesets per method that pass the given FDR.
In this case you have to pass in both a `max.p` and `top.n` parameter.

<div class="note">
Note that we are also asking to plot the **t-statistics** in the distributions
below.
</div>

In the report below, the top 10 *significant* geneset results are reported
*per method*:

```{r show-multiGSEA-top10-fdr, results='asis'}
rmd.multiGSEA('tumor-vs-normal-gsea-top10-fdr', mg, plot.stat='t',
              top.n=10, max.p=0.30)
```

# Comparing voom vs edgeR GSEA

Up until now we have run GSEA using voom'd data, one could also perform
`roast`, `camera`, *and* `romer` gene set testing with count data using `edgeR`.

```{r edger-gsea, cache=TRUE}
mge <- multiGSEA(gsd, y, es.design, contrast=2,
                 methods=c(gsea.methods, 'romer'))
```

## Nominal p-value comparison

We can see from the plot below that the analogous `roast` and `camera` runs
using voom/limma vs edgeR are rather comparable, with edgeR giving slihgtly more
optimistic p-values.

```{r voom-edger-pvals}
cmeths <- c('camera', 'roast')
rv <- results(mg, cmeths)
re <- results(mge, cmeths)
gg <- local({
  cam <- cbind(rv[, list(collection, name, method='camera', pval.v=pval.camera)],
               re[, list(pval.e=pval.camera)])
  ro <- cbind(rv[, list(collection, name, method='roast', pval.v=pval.roast)],
              re[, list(pval.e=pval.roast)])
  rbind(cam, ro)
})

ggplot(gg, aes(-log10(pval.v), -log10(pval.e))) +
  geom_point(alpha=0.8) +
  facet_wrap(~ method) +
  xlab("-log10(voom pvalue)") +
  ylab("-log10(edgeR pvalue)") +
  geom_abline(intercept=0, slope=1, colour='red')
```

## roast vs camera vs romer

Are the pvalues that each method generates comparable to each other at all?
roast is testing a different hypothesis when compared to romer and camera,
so you might suspect that will be an "outlier", but that doesn't really seem
to be the case.

```{r roast-vs-camera-vs-romer}
rea <- results(mge)
comp <- c('pval.roast', 'pval.mixed.roast', 'pval.camera',
          'pval.romer', 'pval.up.romer', 'pval.down.romer')
C <- -log10(as.matrix(rea[, comp, with=FALSE]))
pairs(C, pch=16, col="#3f3f3f33", gap=0.2)
```

# edgeR GSEA results

```{r show-edgeR-multiGSEA, results='asis'}
rmd.multiGSEA('tumor-vs-normal-gsea-edgeR', mge, top.n=10)
```

# PANTHER database

We've also wrapped the [PANTHER database](http://pantherdb.org/) for easy
geneset testing.

```{r edgeR-panther, cache=TRUE}
pdb <- getPanther(type='pathway', species='human')
mgp <- multiGSEA(pdb, y, es.design, contrast=2, 
                 methods=c('camera', 'roast', 'romer'))
```

```{r display-panther-result, results='asis'}
rmd.multiGSEA('tumor-vs-normal-gsea-panther-edgeR', mgp, top.n=20)
```

```
